server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -99,19 +136,27 @@ public Client provide(ContainerRequest containerRequest,
           requestPort, possibleXfccConfig.isEmpty() ? "never" : "only", XFCC_HEADER_NAME));
     }
 
-    // Extract information about the requester. This may be a Keywhiz client, or it may be a proxy
-    // forwarding the real Keywhiz client information in the x-forwarded-client-certs header
-    Principal requestPrincipal = getPrincipal(containerRequest).orElseThrow(
+    // Extract information about the entity that connected directly to Keywhiz.
+    // This must be identified from the request's security context, rather than
+    // easily modified information like a header.
+    //
+    // This entity may be a Keywhiz client, or it may be a proxy
+    // forwarding the real Keywhiz client information.
+    Principal connectedPrincipal = getPrincipal(containerRequest).orElseThrow(
         () -> new NotAuthorizedException("Not authorized as Keywhiz client"));
 
     // Extract client information based on the x-forwarded-client-cert header or
     // on the security context of this request
     if (possibleXfccConfig.isEmpty()) {
-      // The XFCC header is not used; use the security context of this request to identify the client
-      return authenticateClientFromPrincipal(requestPrincipal);
+      // The XFCC header is not used; use the security context of this request to
+      // identify the client
+      return authenticateClientFromPrincipal(connectedPrincipal);
     } else {
-      return authorizeClientFromXfccHeader(possibleXfccConfig.get(), xfccHeaderValues,
-          requestPrincipal, containerRequest);
+      // Verify that the entity that connected directly to Keywhiz is allowed to
+      // forward client data, then use either the XFCC header or a caller-ID header to
+      // identify the client.
+      return authenticateClientFromForwardedData(possibleXfccConfig.get(), xfccHeaderValues,
+          connectedPrincipal, containerRequest);
     }
```

Comment:
Similar to above, I don't think it's a good idea to comment on the callee's non-obvious (i.e. not in the method name) behavior from the caller's code, even if they're in the same class. It looks like the first line of authenticateClientFromForwardedData is a call to validateForwardedClientDataAllowed, so if we need to make that more apparent we could try moving that line just before this one (assuming we're comfortable with allowing those operations to be performed independently).